core/trino-main/src/main/java/io/trino/operator/DeduplicatingDirectExchangeBuffer.java

```diff
@@ -385,29 +405,427 @@ public synchronized void close()
+                ExchangeManagerRegistry exchangeManagerRegistry,
+                QueryId queryId,
+                ExchangeId exchangeId,
+                DataSize pageBufferCapacity)
+        {
+            this.exchangeManagerRegistry = requireNonNull(exchangeManagerRegistry, "exchangeManagerRegistry is null");
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.exchangeId = requireNonNull(exchangeId, "exchangeId is null");
+            this.pageBufferCapacityInBytes = requireNonNull(pageBufferCapacity, "pageBufferCapacity is null").toBytes();
+        }
+
+        public void addPages(TaskId taskId, List<Slice> pages)
+        {
+            if (closed) {
+                return;
+            }
+
+            if (inputFinished) {
+                // ignore extra pages after input is marked as finished
+                return;
+            }
+
+            long pagesRetainedSizeInBytes = getRetainedSizeInBytes(pages);
+            if (exchangeSink == null && pageBufferRetainedSizeInBytes + pagesRetainedSizeInBytes <= pageBufferCapacityInBytes) {
+                pageBuffer.putAll(taskId, pages);
+                pageBufferRetainedSizeInBytes += pagesRetainedSizeInBytes;
+                bufferedPageCount += pages.size();
+                return;
+            }
+
+            if (exchangeSink == null) {
+                verify(exchangeManager == null, "exchangeManager is not expected to be initialized");
+                verify(exchange == null, "exchange is not expected to be initialized");
+                verify(sinkInstanceHandle == null, "sinkInstanceHandle is not expected to be initialized");
+                verify(writeBuffer == null, "writeBuffer is not expected to be initialized");
+
+                try {
+                    exchangeManager = exchangeManagerRegistry.getExchangeManager();
+                }
+                catch (TrinoException e) {
+                    if (EXCHANGE_MANAGER_NOT_CONFIGURED.toErrorCode().equals(e.getErrorCode())) {
+                        throw new TrinoException(EXCHANGE_MANAGER_NOT_CONFIGURED, "Exchange manager must be configured for failure recovery capabilities to be fully functional");
+                    }
+                    throw e;
+                }
+
+                exchange = exchangeManager.createExchange(new ExchangeContext(queryId, exchangeId), 1);
+
+                ExchangeSinkHandle sinkHandle = exchange.addSink(0);
+                sinkInstanceHandle = exchange.instantiateSink(sinkHandle, 0);
+                exchange.noMoreSinks();
+                exchangeSink = exchangeManager.createSink(sinkInstanceHandle, true);
+
+                writeBuffer = new DynamicSliceOutput(DEFAULT_MAX_PAGE_SIZE_IN_BYTES);
+            }
+
+            if (!pageBuffer.isEmpty()) {
+                for (Map.Entry<TaskId, List<Slice>> entry : asMap(pageBuffer).entrySet()) {
+                    writeToSink(entry.getKey(), entry.getValue());
```

Comment:
In theory we can keep them in buffer and don't write to sink. Writing them saves memory and simplifies logic, keeping them will incur reduce the amount of data written to the sink.

Reply:
The sink is likely to have a buffer on it's own. Maintaining two buffers will increase the memory footprint and make the implementation more complex.